File: 백준/Silver/9012. 괄호/괄호.py
import sys
a = []
n = int(sys.stdin.readline())
s = ""
for _ in range(n):
    s = sys.stdin.readline()
    a.append(s[0])
    for i in range(1, len(s)):
        a.append(s[i])
        if a[len(a) - 2] == '(' and a[len(a) - 1] == ')':
            a.pop()
            a.pop()
    a.pop()
    if len(a) == 0:
        print('YES')
    else:
        print('NO')
    a = []

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        
        for (int i = 0; i < t; i++) {
            Stack<Character> st = new Stack<>();
            String s = br.readLine();
            
            for (int k = 0; k < s.length(); k++) {
                char c = s.charAt(k);
                if (!st.isEmpty() && ((st.peek() == '(' && c == ')') || (st.peek() == '[' && c == ']')))
                    st.pop();
                else st.push(c);
            }
            if (st.isEmpty()) System.out.println("YES");
            else System.out.println("NO");
        }
    }
}